uvq_pytorch/utils/contentnet.py

```diff
@@ -0,0 +1,154 @@
+            MBConvSamePadding(192, 6, 192, 5, 1, stochastic_depth_prob[14]),
+            MBConvSamePadding(192, 6, 320, 3, 1, stochastic_depth_prob[15]),
+            Interpolate(size=(16, 16), mode="bilinear", align_corners=False),
+            Conv2dSamePadding(320, 100, kernel_size=16, stride=1),
+        )
+        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
+        self.classifier = nn.Sequential(
+            nn.Dropout(p=dropout),
+            nn.Flatten(),
+            nn.Linear(100, num_classes),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, x):
+        features = self.features(x)
+        x = self.avgpool(features)
+        label_probs = self.classifier(x)
+        return features, label_probs
+
+
+class ContentNetInference:
+    def __init__(
+        self, model_path=MODEL_PATH, num_classes=3862, eval_mode=True, pretrained=True
+    ):
+        self.model = ContentNet(num_classes=num_classes)
+        if pretrained:
+            self.load_state_dict(model_path)
+        if eval_mode:
+            self.model.eval()
+        self.label_mapping: pd.DataFrame = self.load_labels_df(LABELS_CSV_PATH)
+        self.features_transpose = (0, 2, 3, 1)
+
+    def load_state_dict(self, model_path) -> torch.nn.Module:
+        model = torch.load(model_path)
+        self.model.load_state_dict(model)
+        return model
+
+    def predict(self, frame):
+        with torch.no_grad():
+            _, label_probs = self.model(Tensor(np.expand_dims(frame, 0)))
+        return label_probs.detach().numpy()
+
+    def predict_and_get_features(self, frame) -> tuple[np.ndarray, np.ndarray]:
+        with torch.no_grad():
+            features, label_probs = self.model(torch.Tensor(np.expand_dims(frame, 0)))
+        return (
+            features.detach().numpy().transpose(*self.features_transpose),
+            label_probs.detach().numpy()[0],
+        )
+
+    def load_labels_df(self, csv_path) -> pd.DataFrame:
+        df = pd.read_csv(csv_path)
+        return df
+
+    def label_probabilities_to_text(self, label_probs, top_n=1):
+        top_indices = label_probs.argsort()[: -top_n - 1 : -1]
+        probs = label_probs[top_indices]
+        predicted = self.label_mapping.merge(
+            pd.DataFrame({"prob": probs, "Index": top_indices}), on="Index"
+        ).sort_values("prob", ascending=False)
+        return (
+            predicted["Name"].tolist(),
+            predicted["prob"].tolist(),
+            predicted["Index"].tolist(),
+        )
+
+    def infer_from_input_video(self, video_filename, video_length, transpose=False):
+        video, video_resized = VideoReader.load_video(
+            video_filename, video_length, transpose
+        )
+        video_features, video_labels = self.get_labels_and_features_for_all_frames(
+            video_resized
+        )
+        return video_features, video_labels
+
+    def get_labels_and_features_for_all_frames(self, video):
```

Comment:
I would recommend adding a docstring and type annotation for this function (and any other public functions in this file or other files).

For example:
```
def get_labels_and_features_for_all_frames(self, video: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
"""
Gets predicted labels and features for all frames of a video.

Args:
  video: parsed video 5d array with shape (num_seconds, fps, height, width, 3) where each value is in [-1, 1] .

Returns:
  feature: ...
  label: ...

"""
```

Reference: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods